File: 2017/continuations-trampolines/evenodd.py
# Example of mutual recursion with even/odd.
#
# Eli Bendersky [http://eli.thegreenplace.net]
# This code is in the public domain.


def is_even(n):
    if n == 0:
        return True
    else:
        return is_odd(n - 1)


def is_odd(n):
    if n == 0:
        return False
    else:
        return is_even(n - 1)


def is_even_thunked(n):
    if n == 0:
        return True
    else:
        return lambda: is_odd_thunked(n - 1)


def is_odd_thunked(n):
    if n == 0:
        return False
    else:
        return lambda: is_even_thunked(n - 1)


def trampoline(f, *args):
    v = f(*args)
    while callable(v):
        v = v()
    return v


if __name__ == '__main__':
    print(is_even(800))

    # If I try to run is_even(1000) with the default system recursion limit, it
    # blows up. But trampolining keeps the stack depth constant and small.
    print(trampoline(is_even_thunked, 1000))

File: 2017/continuations-trampolines/clojure-samples/src/clojure_samples/evenodd.clj
;;; Example of even/odd in Clojure; recursive and trampolined.
;;;
;;; Eli Bendersky [http://eli.thegreenplace.net]
;;; This code is in the public domain.
(ns clojure-samples.evenodd)

(declare is-even?)

(defn is-odd?
  [n]
  (if (= n 0)
    false
    (is-even? (- n 1))))

(defn is-even?
  [n]
  (if (= n 0)
    true
    (is-odd? (- n 1))))

(map is-even? '(1 2 3 4))

(declare is-even-thunked?)

(defn is-odd-thunked?
  [n]
  (if (= n 0)
    false
    #(is-even-thunked? (- n 1))))

(defn is-even-thunked?
  [n]
  (if (= n 0)
    true
    #(is-odd-thunked? (- n 1))))

(trampoline is-even-thunked? 2)
(trampoline is-even-thunked? 3)
